File: rest-common/src/main/java/org/chronopolis/rest/support/ZonedDateTimeSerializer.java
package org.chronopolis.rest.support;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.lang.reflect.Type;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;

/**
 * ZonedDateTime serializer into an ISO-8601 string (YYYY-MM-ddTHH:mm:ssZ)
 *
 * Created by shake on 3/22/16.
 */
public class ZonedDateTimeSerializer implements JsonSerializer<ZonedDateTime> {
    @Override
    public JsonElement serialize(ZonedDateTime zonedDateTime, Type type, JsonSerializationContext jsonSerializationContext) {
        DateTimeFormatter fmt = DateTimeFormatter.ISO_OFFSET_DATE_TIME.withZone(ZoneOffset.UTC);
        return new JsonPrimitive(fmt.format(zonedDateTime.truncatedTo(ChronoUnit.SECONDS)));
    }
}

File: rest-models/src/main/kotlin/org/chronopolis/rest/kot/models/serializers/ZonedDateTimeSerializer.kt
package org.chronopolis.rest.kot.models.serializers

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.databind.JsonSerializer
import com.fasterxml.jackson.databind.SerializerProvider
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter.ISO_OFFSET_DATE_TIME

/**
 * Simple serializer for writing [ZonedDateTime] as an ISO-8601 string at UTC-0
 *
 * @author shake
 */
class ZonedDateTimeSerializer : JsonSerializer<ZonedDateTime>() {
    override fun serialize(dateTime: ZonedDateTime,
                           jsonGenerator: JsonGenerator,
                           serializerProvider: SerializerProvider) {
        jsonGenerator.writeString(ISO_OFFSET_DATE_TIME.withZone(ZoneOffset.UTC).format(dateTime))
    }
}

